processing/src/main/java/io/druid/segment/StringDimensionHandler.java

```diff
@@ -0,0 +1,1248 @@
+    }
+
+    public String getValue(int id)
+    {
+      synchronized (lock) {
+        return Strings.emptyToNull(idToValue.get(id));
+      }
+    }
+
+    public boolean contains(String value)
+    {
+      synchronized (lock) {
+        return valueToId.containsKey(value);
+      }
+    }
+
+    public int size()
+    {
+      synchronized (lock) {
+        return valueToId.size();
+      }
+    }
+
+    public int add(String originalValue)
+    {
+      String value = Strings.nullToEmpty(originalValue);
+      synchronized (lock) {
+        Integer prev = valueToId.get(value);
+        if (prev != null) {
+          return prev;
+        }
+        final int index = size();
+        valueToId.put(value, index);
+        idToValue.add(value);
+        minValue = minValue == null || minValue.compareTo(value) > 0 ? value : minValue;
+        maxValue = maxValue == null || maxValue.compareTo(value) < 0 ? value : maxValue;
+        return index;
+      }
+    }
+
+    public String getMinValue()
+    {
+      return minValue;
+    }
+
+    public String getMaxValue()
+    {
+      return maxValue;
+    }
+
+    public SortedDimensionDictionary sort()
+    {
+      synchronized (lock) {
+        return new SortedDimensionDictionary(idToValue, size());
+      }
+    }
+  }
+
+  private class SortedDimensionDictionary
+  {
+    private final List<String> sortedVals;
+    private final int[] idToIndex;
+    private final int[] indexToId;
+
+    public SortedDimensionDictionary(List<String> idToValue, int length)
+    {
+      Map<String, Integer> sortedMap = Maps.newTreeMap();
+      for (int id = 0; id < length; id++) {
+        sortedMap.put(idToValue.get(id), id);
+      }
+      this.sortedVals = Lists.newArrayList(sortedMap.keySet());
+      this.idToIndex = new int[length];
+      this.indexToId = new int[length];
+      int index = 0;
+      for (Integer id : sortedMap.values()) {
+        idToIndex[id] = index;
+        indexToId[index] = id;
+        index++;
+      }
+    }
+
+    public int size()
+    {
+      return sortedVals.size();
+    }
+
+    public int getUnsortedIdFromSortedId(int index)
+    {
+      return indexToId[index];
+    }
+
+    public int getSortedIdFromUnsortedId(int id)
+    {
+      return idToIndex[id];
+    }
+
+    public String getValueFromSortedId(int index)
+    {
+      return Strings.emptyToNull(sortedVals.get(index));
+    }
+  }
+
+  public class StringDimensionIndexer implements DimensionIndexer<Integer, String>
+  {
+    private DimensionDictionary dimLookup;
+    private SortedDimensionDictionary sortedLookup;
+    private Object lock;
+
+    public StringDimensionIndexer(Object lock)
+    {
+      this.lock = lock;
+      this.dimLookup = new DimensionDictionary(this.lock);
+    }
+
+    @Override
+    public void addNullLookup()
+    {
+      int oldDictSize = dimLookup.size();
+      dimLookup.add(null);
+      if (oldDictSize != dimLookup.size()) {
+        sortedLookup = dimLookup.sort();
```

Comment:
can we just set sortedLoojup to null here and let it be created when needed by something ?